File: packages/compose/src/main/kotlin/ai/cocso/ui/CCAvatar.kt
package ai.cocso.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** An identity mark: initials when there is no image. */
@Composable
fun CCAvatar(
    initials: String,
    label: String,
    modifier: Modifier = Modifier,
    size: CCAvatarSize = CCAvatarSize.md,
    shape: CCAvatarShape = CCAvatarShape.circle,
) {
    val style = cCAvatarStyle(size = size, shape = shape)
    val side = style.width ?: 32.dp
    val clipShape =
        // A percentage radius in the recipe arrives as `borderRadiusFull`.
        // Reading the shape off the variant name instead is how the two
        // platforms ended up compensating in two different ways.
        if (style.borderRadiusFull == true) CircleShape
        else RoundedCornerShape(style.borderRadius ?: 0.dp)

    Box(
        modifier = modifier
            .size(side, style.height ?: side)
            .clip(clipShape)
            .background(style.bgColor ?: CocsoTokens.Color.surfaceNeutral())
            // Named once, so a screen reader does not announce an unlabelled
            // element beside the label.
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = initials,
            color = style.fontColor ?: CocsoTokens.Color.textPrimary(),
            fontSize = (style.fontSize?.value ?: 12f).sp,
            fontWeight = style.fontWeight ?: FontWeight.SemiBold,
        )
    }
}

File: packages/swiftui/Sources/CocsoUI/CCAvatar.swift
import SwiftUI

/// An identity mark: an image, or initials when there is none.
public struct CCAvatar: View {
    private let initials: String
    private let image: Image?
    private let size: CCAvatarSize
    private let shape: CCAvatarShape
    private let label: String

    @Environment(\.colorScheme) private var colorScheme
    // The brand the app set at its root. Tokens and resolvers take it so a
    // design-system view draws the same primary the app does.
    @Environment(\.cocsoBrand) private var brand

    public init(
        initials: String,
        label: String,
        image: Image? = nil,
        size: CCAvatarSize = .md,
        shape: CCAvatarShape = .circle
    ) {
        self.initials = initials
        self.label = label
        self.image = image
        self.size = size
        self.shape = shape
    }

    public var body: some View {
        let style = CCAvatarStyle.resolve(size: size, shape: shape, scheme: colorScheme, brand: brand)
        let side = style.width ?? 32
        Group {
            if let image {
                image.resizable().scaledToFill()
            } else {
                Text(initials)
                    .font(.system(size: style.fontSize ?? 12, weight: style.fontWeight ?? .semibold))
                    .foregroundStyle(style.fontColor ?? CocsoTokens.Color.textPrimary(colorScheme, brand: brand))
            }
        }
        .frame(width: side, height: style.height ?? side)
        .background(style.bgColor ?? CocsoTokens.Color.surfaceNeutral(colorScheme, brand: brand))
        .clipShape(
            style.borderRadiusFull == true
                ? AnyShape(Circle())
                : AnyShape(RoundedRectangle(cornerRadius: style.borderRadius ?? 0))
        )
        // The image carries no meaning the label does not; naming the element
        // once keeps a screen reader from announcing an unlabelled image.
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label)
    }
}
